[Bug] [0.7.0]使用官网自定义udf时，报错implementation errors，ClassNotFoundException

### Search before asking

- [X] I had searched in the [issues](https://github.com/DataLinkDC/dinky/issues?q=is%3Aissue) and found no similar issues.


### What happened

使用官网http://www.dlink.top/docs/0.7/udf_develop/how_to/java&scala_udf的模板udf代码报错

代码如下
![image](https://user-images.githubusercontent.com/123799101/215253811-f98dc1f8-e09e-4584-a27e-2b6bf5e4226c.png)

![image](https://user-images.githubusercontent.com/123799101/215254213-117193ce-a611-4983-b3e0-2cc5a405d4d9.png)


校验报错如下

![image](https://user-images.githubusercontent.com/123799101/215254048-0bdfdaa2-e616-44b8-95ae-d4eb94ae13e7.png)

具体报错内容如下

org.apache.flink.table.api.ValidationException: Could not register temporary catalog function 'default_catalog.default_database.sb_j' due to implementation errors.
	at org.apache.flink.table.catalog.FunctionCatalog.registerTemporaryCatalogFunction(FunctionCatalog.java:160)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.createCatalogFunction(TableEnvironmentImpl.java:1824)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:1158)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeSql(TableEnvironmentImpl.java:730)
	at com.dlink.executor.Executor.executeSql(Executor.java:247)
	at com.dlink.explainer.Explainer.explainSql(Explainer.java:232)
	at com.dlink.job.JobManager.explainSql(JobManager.java:669)
	at com.dlink.service.impl.StudioServiceImpl.explainFlinkSql(StudioServiceImpl.java:272)
	at com.dlink.service.impl.StudioServiceImpl.explainSql(StudioServiceImpl.java:257)
	at com.dlink.service.impl.StudioServiceImpl$$FastClassBySpringCGLIB$$e3eb787.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:89)
	at com.dlink.aop.UdfClassLoaderAspect.round(UdfClassLoaderAspect.java:65)
	at sun.reflect.GeneratedMethodAccessor117.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:634)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:624)
	at org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:72)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
	at com.dlink.service.impl.StudioServiceImpl$$EnhancerBySpringCGLIB$$96eb6e4b.explainSql(<generated>)
	at com.dlink.controller.StudioController.explainSql(StudioController.java:87)
	at com.dlink.controller.StudioController$$FastClassBySpringCGLIB$$e6483d87.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.framework.adapter.AfterReturningAdviceInterceptor.invoke(AfterReturningAdviceInterceptor.java:57)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
	at com.dlink.controller.StudioController$$EnhancerBySpringCGLIB$$288fbb9d.explainSql(<generated>)
	at sun.reflect.GeneratedMethodAccessor344.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1071)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:696)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:779)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at com.alibaba.druid.support.http.WebStatFilter.doFilter(WebStatFilter.java:124)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:893)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1789)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: com.test.SubFunction
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at org.apache.flink.util.FlinkUserCodeClassLoader.loadClassWithoutExceptionHandling(FlinkUserCodeClassLoader.java:67)
	at org.apache.flink.util.ChildFirstClassLoader.loadClassWithoutExceptionHandling(ChildFirstClassLoader.java:74)
	at org.apache.flink.util.FlinkUserCodeClassLoader.loadClass(FlinkUserCodeClassLoader.java:51)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at org.apache.flink.util.FlinkUserCodeClassLoaders$SafetyNetWrapperClassLoader.loadClass(FlinkUserCodeClassLoaders.java:192)
	at org.apache.flink.table.catalog.FunctionCatalog.validateAndPrepareFunction(FunctionCatalog.java:680)
	at org.apache.flink.table.catalog.FunctionCatalog.registerTemporaryCatalogFunction(FunctionCatalog.java:157)
	... 94 more


### What you expected to happen

找不到对应接口的实现类，但是已经按照最简单的模板去写了对应的实现类，是不是什么原因，写的java代码并没有加载到ScalarFunction类的实现下面。具体原因不明。

### How to reproduce

就是官网的代码基本没变，不同的主要是环境，是不是由于环境的问题？
以下是使用的环境

![image](https://user-images.githubusercontent.com/123799101/215254476-d1c415fa-73a6-4677-8ff9-16e2b0beb319.png)


![image](https://user-images.githubusercontent.com/123799101/215254514-67dd09b5-abc2-4a7c-b15c-b9f5cb65726d.png)


### Anything else

_No response_

### Version

0.7.0

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/dinky-client/dinky-client-1.16/src/main/java/org/dinky/executor/CustomTableEnvironmentImpl.java
+++ b/dinky-client/dinky-client-1.16/src/main/java/org/dinky/executor/CustomTableEnvironmentImpl.java
@@ -22,6 +22,7 @@ package org.dinky.executor;
 import static org.apache.flink.table.api.bridge.internal.AbstractStreamTableEnvironmentImpl.lookupExecutor;
 
 import org.dinky.assertion.Asserts;
+import org.dinky.context.DinkyClassLoaderContextHolder;
 import org.dinky.model.LineageRel;
 import org.dinky.result.SqlExplainResult;
 import org.dinky.utils.FlinkStreamProgramWithoutPhysical;
@@ -129,7 +130,9 @@ public class CustomTableEnvironmentImpl extends AbstractCustomTableEnvironment {
             StreamExecutionEnvironment executionEnvironment, EnvironmentSettings settings) {
         final MutableURLClassLoader userClassLoader =
                 FlinkUserCodeClassLoaders.create(
-                        new URL[0], settings.getUserClassLoader(), settings.getConfiguration());
+                        new URL[0],
+                        DinkyClassLoaderContextHolder.get(),
+                        settings.getConfiguration());
         final Executor executor = lookupExecutor(userClassLoader, executionEnvironment);
 
         final TableConfig tableConfig = TableConfig.getDefault();

--- a/dinky-common/src/main/java/org/dinky/classloader/DinkyClassLoader.java
+++ b/dinky-common/src/main/java/org/dinky/classloader/DinkyClassLoader.java
@@ -44,18 +44,7 @@ public class DinkyClassLoader extends URLClassLoader {
 
     public DinkyClassLoader(Collection<File> fileSet, ClassLoader parent) {
         super(new URL[] {}, parent);
-        URL[] urls =
-                fileSet.stream()
-                        .map(
-                                x -> {
-                                    try {
-                                        return x.toURI().toURL();
-                                    } catch (MalformedURLException e) {
-                                        throw new RuntimeException(e);
-                                    }
-                                })
-                        .toArray(URL[]::new);
-        addURL(urls);
+        addURL(fileSet);
     }
 
     public DinkyClassLoader(URL[] urls) {
@@ -70,6 +59,21 @@ public class DinkyClassLoader extends URLClassLoader {
         for (URL url : urls) {
             super.addURL(url);
         }
+    }
+
+    public void addURL(Collection<File> fileSet) {
+        URL[] urls =
+                fileSet.stream()
+                        .map(
+                                x -> {
+                                    try {
+                                        return x.toURI().toURL();
+                                    } catch (MalformedURLException e) {
+                                        throw new RuntimeException(e);
+                                    }
+                                })
+                        .toArray(URL[]::new);
+        addURL(urls);
     }
 
     public void addURL(String[] paths, List<String> notExistsFiles) {

--- a/dinky-common/src/main/java/org/dinky/context/DinkyClassLoaderContextHolder.java
+++ b/dinky-common/src/main/java/org/dinky/context/DinkyClassLoaderContextHolder.java
@@ -22,6 +22,7 @@ package org.dinky.context;
 import org.dinky.classloader.DinkyClassLoader;
 
 import java.io.IOException;
+import java.net.URL;
 
 import lombok.extern.slf4j.Slf4j;
 
@@ -35,14 +36,19 @@ public class DinkyClassLoaderContextHolder {
     private static final ThreadLocal<DinkyClassLoader> CLASS_LOADER_CONTEXT = new ThreadLocal<>();
     private static final ThreadLocal<ClassLoader> INIT_CLASS_LOADER_CONTEXT = new ThreadLocal<>();
 
-    public static void set(DinkyClassLoader classLoader) {
+    protected static void set(DinkyClassLoader classLoader) {
         CLASS_LOADER_CONTEXT.set(classLoader);
         INIT_CLASS_LOADER_CONTEXT.set(Thread.currentThread().getContextClassLoader());
         Thread.currentThread().setContextClassLoader(classLoader);
     }
 
     public static DinkyClassLoader get() {
-        return CLASS_LOADER_CONTEXT.get();
+        DinkyClassLoader dinkyClassLoader = CLASS_LOADER_CONTEXT.get();
+        if (dinkyClassLoader == null) {
+            dinkyClassLoader = new DinkyClassLoader(new URL[0]);
+            set(dinkyClassLoader);
+        }
+        return dinkyClassLoader;
     }
 
     public static void clear() {

--- a/dinky-core/src/main/java/org/dinky/job/JobManager.java
+++ b/dinky-core/src/main/java/org/dinky/job/JobManager.java
@@ -25,7 +25,6 @@ import static org.dinky.function.util.UDFUtil.YARN;
 
 import org.dinky.api.FlinkAPI;
 import org.dinky.assertion.Asserts;
-import org.dinky.classloader.DinkyClassLoader;
 import org.dinky.constant.FlinkSQLConstant;
 import org.dinky.context.DinkyClassLoaderContextHolder;
 import org.dinky.context.JarPathContextHolder;
@@ -418,13 +417,11 @@ public class JobManager {
             }
         }
 
-        DinkyClassLoader classLoader =
-                new DinkyClassLoader(
+        DinkyClassLoaderContextHolder.get()
+                .addURL(
                         CollUtil.addAll(
                                 JarPathContextHolder.getUdfFile(),
-                                JarPathContextHolder.getOtherPluginsFiles()),
-                        Thread.currentThread().getContextClassLoader());
-        DinkyClassLoaderContextHolder.set(classLoader);
+                                JarPathContextHolder.getOtherPluginsFiles()));
     }
 
     public JobResult executeSql(String statement) {

